### PR TITLE
Ck tile batched gemm example

### PR DESCRIPTION
**Pull request by aledudek**

### CODE REVIEW

File: example/ck_tile/12_batched_gemm/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+add_executable(tile_example_batched_gemm EXCLUDE_FROM_ALL batched_gemm.cpp)
```
Comment (bartekxk) [inline example/ck_tile/12_batched_gemm/CMakeLists.txt]: please add new line at the end

File: example/ck_tile/12_batched_gemm/run_batched_gemm_example.inc [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    args.batch_stride_C = batch_stride_C;
+    args.batch_count    = batch_count;
+
+    float ave_time = gemm_calc<ALayout, BLayout, CLayout>(
+        args, ck_tile::stream_config{nullptr, true, 1, n_warmup, n_repeat});
+
+    std::string op_name{"Gemm{MemBoundPipeline}"};
+
+    std::size_t flop = std::size_t(2) * M * N * K;
+    std::size_t num_byte =
```
Comment (bartekxk) [inline example/ck_tile/12_batched_gemm/run_batched_gemm_example.inc]: please update num_byte and flop formulas. You have to include batch count

File: example/ck_tile/12_batched_gemm/run_batched_gemm_example.inc [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    using namespace ck_tile::literals;
+
+    auto f_host_tensor_descriptor = [](std::size_t batch_count_,
+                                       std::size_t row,
+                                       std::size_t col,
+                                       std::size_t stride,
+                                       auto layout) {
+        if constexpr(std::is_same_v<decltype(layout), ck_tile::tensor_layout::gemm::RowMajor>)
+        {
+            return ck_tile::HostTensorDescriptor({batch_count_, row, col},
```
Comment (bartekxk) [inline example/ck_tile/12_batched_gemm/run_batched_gemm_example.inc]: maybe replace row * col with batch_stride

File: include/ck_tile/host/reference/reference_gemm.hpp [hunk 4]
```diff
@@ -183,4 +183,116 @@ void reference_gemm_gpu(DeviceMem& a_device,
...
+    if(errB != hipSuccess)
+    {
+        std::cerr << "Error copying B to device: " << hipGetErrorString(errB) << std::endl;
+    }
+
+    int totalElements      = M * N;
+    int numThreadsPerBlock = 256; // Common choice for threads per block
+    int numBlocks          = (totalElements + numThreadsPerBlock - 1) / numThreadsPerBlock;
+
+    for(int i = 0; i < batch_count; ++i)
```
Comment (bartekxk) [inline include/ck_tile/host/reference/reference_gemm.hpp]: maybe change `int i` to `index_t batch_id`

File: include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp [hunk 5]
```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2024, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (bartekxk) [inline include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp]:
```suggestion
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
```

File: include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    using GemmPipeline                       = remove_cvref_t<GemmPipeline_>;
+    using EpiloguePipeline                   = remove_cvref_t<EpiloguePipeline_>;
+    using ALayout                            = remove_cvref_t<typename GemmPipeline::ALayout>;
+    using BLayout                            = remove_cvref_t<typename GemmPipeline::BLayout>;
+    using CLayout                            = remove_cvref_t<typename GemmPipeline::CLayout>;
+    static constexpr index_t KernelBlockSize = GemmPipeline::BlockSize;
+
+    using ADataType = remove_cvref_t<typename GemmPipeline::ADataType>;
+    using BDataType = remove_cvref_t<typename GemmPipeline::BDataType>;
+    // using CAccDataType = remove_cvref_t<typename GemmPipeline::CDataType>;
```
Comment (bartekxk) [inline include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp]:
```suggestion
```

File: include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+    CK_TILE_HOST_DEVICE static constexpr index_t GetSmemSize()
+    {
+        return max(GemmPipeline::GetSmemSize(), EpiloguePipeline::GetSmemSize());
+    }
+
+    CK_TILE_DEVICE void operator()(BatchedGemmCommonKargs kargs) const
+    {
+        const auto [i_m, i_n] = TilePartitioner{}();
+        const auto i_k        = blockIdx.z;
```
Comment (bartekxk) [inline include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp]: please change `i_k` to `i_batch`

File: include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+    using BLayout                            = remove_cvref_t<typename GemmPipeline::BLayout>;
+    using CLayout                            = remove_cvref_t<typename GemmPipeline::CLayout>;
+    static constexpr index_t KernelBlockSize = GemmPipeline::BlockSize;
+
+    using ADataType = remove_cvref_t<typename GemmPipeline::ADataType>;
+    using BDataType = remove_cvref_t<typename GemmPipeline::BDataType>;
+    // using CAccDataType = remove_cvref_t<typename GemmPipeline::CDataType>;
+    using CDataType = remove_cvref_t<typename EpiloguePipeline::ODataType>;
+
+    __host__ static constexpr auto GridSize(index_t M, index_t N, index_t KBatch)
```
Comment (bartekxk) [inline include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp]: Could you add fourth parameter `BatchCount`? Then you can multiply KBatch times BatchCount. Add please add some check like this:
`if (KBatch > 1){ std::runtime_error("split k is not supported");}`

File: include/ck_tile/host/reference/reference_gemm.hpp [hunk 9]
```diff
@@ -183,4 +183,116 @@ void reference_gemm_gpu(DeviceMem& a_device,
...
+                                index_t batch_stride_B,
+                                index_t batch_stride_C,
+                                index_t batch_count)
+{
+
+    ADataType* d_A;
+    BDataType* d_B;
+    CDataType* d_C;
+
+    hipError_t errA = hipMalloc(&d_A, batch_count * M * K * sizeof(ADataType));
```
Comment (carlushuang) [inline include/ck_tile/host/reference/reference_gemm.hpp:213]: I saw hipmalloc/memcpy inside gpu reference gemm. Can we remove this and allow outside caller to manage  the gpu buffer?
The reason is this function could be a fallback gpu function in the future, so we could have no malloc/memcpy inside this function (usually gpu memory management is by torch)

Comment (aledudek) [reply include/ck_tile/host/reference/reference_gemm.hpp:213]: Yes, I will apply this change in a separate PR

**Review by bartekxk — DISMISSED** [review]
LGTM

File: example/ck_tile/13_batched_gemm/README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+# you can replace <arch> with the appropriate architecture (for example gfx90a or gfx942) or leave it blank
+sh ../script/cmake-ck-dev.sh  ../ <arch>
+make tile_example_batched_gemm -j
+```
+This will result in an executable `build/bin/tile_example_batched_gemm`
+
+## example
+```
+args:
+        -k_batch     Batch Size (default:1)
```
Comment (aosewski) [inline example/ck_tile/13_batched_gemm/README.md]: We currently do not support Splitk so please remove this.

File: example/ck_tile/13_batched_gemm/batched_gemm.cpp [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+    using CodegenGemmPipeline = ck_tile::GemmPipelineAGmemBGmemCRegV1<CodegenPipelineProblem>;
+    // ToDo: Will add the codegen part to test different pipeline policies in GEMM.
+    // Now we only use the BlockGemmASmemBSmemCRegV1DefaultPolicy.
+    using Kernel = ck_tile::BatchedGemmKernel<TilePartitioner, CodegenGemmPipeline, GemmEpilogue>;
+
+    auto kargs = Kernel::MakeKargs(args.p_a,
+                                   args.p_b,
+                                   args.p_c,
+                                   args.k_batch,
```
Comment (aosewski) [inline example/ck_tile/13_batched_gemm/batched_gemm.cpp]: We do not yet support splitkK so please do not expose this parameter.

File: example/ck_tile/13_batched_gemm/batched_gemm.cpp [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+                                   args.K,
+                                   args.stride_A,
+                                   args.stride_B,
+                                   args.stride_C,
+                                   args.batch_stride_A,
+                                   args.batch_stride_B,
+                                   args.batch_stride_C,
+                                   args.batch_count);
+
+    const dim3 grids      = Kernel::GridSize(args.M, args.N, args.k_batch, args.batch_count);
```
Comment (aosewski) [inline example/ck_tile/13_batched_gemm/batched_gemm.cpp]: We do not yet support splitkK so please do not use k_batch parameter.

File: example/ck_tile/13_batched_gemm/batched_gemm.hpp [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+struct BatchedGemmTypeConfig;
+
+template <>
+struct BatchedGemmTypeConfig<ck_tile::half_t>
+{
+    using ADataType   = ck_tile::half_t;
+    using BDataType   = ck_tile::half_t;
+    using AccDataType = float;
+    using CDataType   = ck_tile::half_t;
+    // ToDo: Add more bias config to support different categories of GEMM.
```
Comment (aosewski) [inline example/ck_tile/13_batched_gemm/batched_gemm.hpp]: Please remove this comment.

File: example/ck_tile/13_batched_gemm/batched_gemm.hpp [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+struct DataTypeTraits<double>
+{
+    static constexpr const char* name = "fp64";
+};
+
+template <>
+struct DataTypeTraits<ck_tile::half_t>
+{
+    static constexpr const char* name = "fp16";
+};
```
Comment (aosewski) [inline example/ck_tile/13_batched_gemm/batched_gemm.hpp]: Are you using this? This is the same as for other gemm examples. Please do not duplicate.

File: example/ck_tile/13_batched_gemm/batched_gemm.hpp [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+using BDataType   = Types::BDataType;
+using AccDataType = Types::AccDataType;
+using CDataType   = Types::CDataType;
+
+struct batched_gemm_args
+{
+    const void* p_a;
+    const void* p_b;
+    void* p_c;
+    ck_tile::index_t k_batch;
```
Comment (aosewski) [inline example/ck_tile/13_batched_gemm/batched_gemm.hpp]: We do not support SplitK yet, thus please remove this param.

File: example/ck_tile/13_batched_gemm/batched_gemm.hpp [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+using Types = BatchedGemmTypeConfig<ck_tile::half_t>;
+
+// Specific type aliases for easy access
+using ADataType   = Types::ADataType;
+using BDataType   = Types::BDataType;
+using AccDataType = Types::AccDataType;
+using CDataType   = Types::CDataType;
+
+struct batched_gemm_args
```
Comment (aosewski) [inline example/ck_tile/13_batched_gemm/batched_gemm.hpp]: Take a look here: https://github.com/ROCm/composable_kernel/blob/develop/include/ck_tile/ops/topk_softmax/kernel/topk_softmax_kernel.hpp#L15-L25
I think that this is a good idea to place host args in kernel file.

Comment (aosewski) [reply example/ck_tile/13_batched_gemm/batched_gemm.hpp]: Then here you just do like this: https://github.com/ROCm/composable_kernel/blob/develop/example/ck_tile/09_topk_softmax/topk_softmax_api.hpp#L17-L19

File: example/ck_tile/13_batched_gemm/batched_gemm.hpp [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+    ck_tile::index_t batch_stride_A;
+    ck_tile::index_t batch_stride_B;
+    ck_tile::index_t batch_stride_C;
+    ck_tile::index_t batch_count;
+};
+
+auto create_args(int argc, char* argv[])
+{
+    ck_tile::ArgParser arg_parser;
+    arg_parser.insert("k_batch", "1", "Batch Size")
```
Comment (aosewski) [inline example/ck_tile/13_batched_gemm/batched_gemm.hpp]: We need also layout for each input tensor.

File: example/ck_tile/13_batched_gemm/run_batched_gemm_example.inc [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+    ck_tile::index_t batch_stride_B = arg_parser.get_int("batch_stride_b");
+    ck_tile::index_t batch_stride_C = arg_parser.get_int("batch_stride_c");
+    ck_tile::index_t batch_count    = arg_parser.get_int("batch_count");
+
+    int n_warmup = arg_parser.get_int("warmup");
+    int n_repeat = arg_parser.get_int("repeat");
+
+    using ALayout = ck_tile::tensor_layout::gemm::RowMajor;
+    using BLayout = ck_tile::tensor_layout::gemm::ColumnMajor;
+    using CLayout = ck_tile::tensor_layout::gemm::RowMajor;
```
Comment (aosewski) [inline example/ck_tile/13_batched_gemm/run_batched_gemm_example.inc]: The layout could be defined in run time by the user. Take a look in here: https://github.com/ROCm/composable_kernel/blob/develop/example/ck_tile/03_gemm/run_gemm_example.inc

File: include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
+
+#pragma once
+
+#include <iostream>
+#include <string>
+
+#include "ck_tile/core.hpp"
+#include "ck_tile/ops/common.hpp"
+#include "ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_scheduler.hpp"
```
Comment (aosewski) [inline include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp]: Rather unused include.

File: include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+    using BDataType = remove_cvref_t<typename GemmPipeline::BDataType>;
+    using CDataType = remove_cvref_t<typename EpiloguePipeline::ODataType>;
+
+    __host__ static constexpr auto
+    GridSize(index_t M, index_t N, index_t KBatch, index_t BatchCount)
+    {
+        if(KBatch > 1)
+        {
+            std::runtime_error("split k is not supported");
+        }
```
Comment (aosewski) [inline include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp]: I'd rather go with not exposing params which we do not support yet.

File: include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+        return TilePartitioner::GridSize(M, N, KBatch * BatchCount);
+    }
+
+    __host__ static constexpr auto BlockSize() { return dim3(KernelBlockSize); }
+
+    struct BatchedGemmCommonKargs
```
Comment (aosewski) [inline include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp]:
```suggestion
    struct BatchedGemmKargs
```

File: include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+        return max(GemmPipeline::GetSmemSize(), EpiloguePipeline::GetSmemSize());
+    }
+
+    CK_TILE_DEVICE void operator()(BatchedGemmCommonKargs kargs) const
+    {
+        const auto [i_m, i_n] = TilePartitioner{}();
+        const auto i_batch    = blockIdx.z;
+        //  options
+        const ADataType* a_start = static_cast<const ADataType*>(kargs.a_ptr) +
+                                   __builtin_amdgcn_readfirstlane(i_batch * kargs.batch_stride_A);
```
Comment (aosewski) [inline include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp]: This readfirstlane should be "outside" that is cover result of a_ptr + ibatch* stride

File: include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp [hunk 23]
```diff
@@ -0,0 +1,220 @@
...
+                    number<GemmPipeline::VectorSizeA>{},
+                    number<1>{});
+            }
+            else
+            {
+                return make_naive_tensor_view<address_space_enum::global>(
+                    a_start,
+                    make_tuple(kargs.M, kargs.K),
+                    make_tuple(1, kargs.stride_A),
+                    number<GemmPipeline::VectorSizeA>{},
```
Comment (aosewski) [inline include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp]: This is a bug, you can't use vector read on a not contiguous dimension. Take a look in here: https://github.com/ROCm/composable_kernel/blob/develop/include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp#L77-L95

File: include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp [hunk 24]
```diff
@@ -0,0 +1,220 @@
...
+                    number<GemmPipeline::VectorSizeB>{},
+                    number<1>{});
+            }
+            else
+            {
+                return make_naive_tensor_view<address_space_enum::global>(
+                    b_start,
+                    make_tuple(kargs.N, kargs.K),
+                    make_tuple(kargs.stride_B, 1),
+                    number<GemmPipeline::VectorSizeB>{},
```
Comment (aosewski) [inline include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:149]: As above

File: include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp [hunk 25]
```diff
@@ -0,0 +1,220 @@
...
+        __shared__ char smem_ptr[GetSmemSize()];
+
+        const index_t num_loop = TilePartitioner::GetLoopNum(kargs.K);
+
+        // Run GEMM cooperatively by whole wokrgroup.
+        auto c_block_tile =
+            GemmPipeline{}.template operator()(a_block_window, b_block_window, num_loop, smem_ptr);
+
+        CDataType* c_start = static_cast<CDataType*>(kargs.c_ptr) +
+                             __builtin_amdgcn_readfirstlane(i_batch * kargs.batch_stride_C);
```
Comment (aosewski) [inline include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp]: as above

File: include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp [hunk 26]
```diff
@@ -0,0 +1,220 @@
...
+                    number<GemmPipeline::VectorSizeC>{},
+                    number<1>{});
+            }
+            else
+            {
+                return make_naive_tensor_view<address_space_enum::global>(
+                    c_start,
+                    make_tuple(kargs.M, kargs.N),
+                    make_tuple(1, kargs.stride_C),
+                    number<GemmPipeline::VectorSizeC>{},
```
Comment (aosewski) [inline include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp]: As above

**Review by aosewski — CHANGES_REQUESTED** [review]
Good start! But need to polish code a bit. Please also write unit tests.

File: example/ck_tile/16_batched_gemm/batched_gemm.hpp [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+        .insert("warmup", "50", "number of iterations before benchmark the kernel")
+        .insert("repeat", "100", "number of iterations to benchmark the kernel")
+        .insert("timer", "gpu", "gpu:gpu timer, cpu:cpu timer");
+
+    bool result = arg_parser.parse(argc, argv);
+    return std::make_tuple(result, arg_parser);
+}
+
+// host API
+float gemm_calc(batched_gemm_kargs args, const ck_tile::stream_config& s);
```
Comment (aosewski) [inline example/ck_tile/16_batched_gemm/batched_gemm.hpp]:
```suggestion
float batched_gemm(batched_gemm_kargs args, const ck_tile::stream_config& s);
```

File: include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp [hunk 28]
```diff
@@ -0,0 +1,227 @@
...
+                    number<1>{});
+            }
+        }();
+
+        auto a_pad_view = pad_tensor_view(
+            a_tensor_view,
+            make_tuple(number<TilePartitioner::kM>{}, number<TilePartitioner::kK>{}),
+            // somehow clang-format is splitting below line into multiple.
+            // clang-format off
+            sequence<false, GemmPipeline::kPadM>{});
```
Comment (aosewski) [inline include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp]: Take a look at gemm_kernel how its done righ tnow.

File: include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp [hunk 29]
```diff
@@ -0,0 +1,227 @@
...
+        // clang-format on
+
+        auto a_block_window = make_tile_window(
+            a_pad_view,
+            make_tuple(number<TilePartitioner::kM>{}, number<TilePartitioner::kK>{}),
+            {i_m, 0});
+
+        auto b_pad_view = pad_tensor_view(
```
Comment (aosewski) [inline include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp]: Take a look at gemm_kernel how its done righ tnow.

File: include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp [hunk 30]
```diff
@@ -0,0 +1,227 @@
...
+                return make_naive_tensor_view<address_space_enum::global>(
+                    c_start + batch_offset_C,
+                    make_tuple(kargs.M, kargs.N),
+                    make_tuple(1, kargs.stride_C),
+                    number<1>{},
+                    number<1>{});
+            }
+        }();
+
+        auto c_pad_view = pad_tensor_view(
```
Comment (aosewski) [inline include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp]: Take a look at gemm_kernel how its done righ tnow.

**Review by aosewski — CHANGES_REQUESTED** [review]
Please catch-up with latest gemm-kernel.

**Review by aosewski — DISMISSED** [review]

**Comment by aosewski** [comment]
Let's wait for CI to pass.

**Review by bartekxk — APPROVED** [review]

**Pull request merged**